Improve performance and add reflink support

kakra opened the pull request:
This branch adds a few cleanups first to fix indentation and an issue reported by the compiler. It then adds a series of patches which try to improve performance.

The most important patch is support for the FICLONE ioctl which yields a performance boost of 50% on filesystems supporting it. No thorough testing has been done, it may eat your kitten. Please do your own checks. I tested this against some big files multiple times comparing md5sums before and after and it reported no data damage, so it seems to work.

This builds upon my integration branch.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/unbrice/shake/14)
<!-- Reviewable:end -->

kakra committed:
Fix calculation of last file ideal position

I think this is a bug so I fix this: It seems that this means to
calculate the ideal position of this file just before the last file in
the order. To do this correctly, we need to subtract y->size and not
y->blocks, as z->start is in bytes and not in blocks.

kakra committed:
Provide better filesystem hinting

The hinting currently doesn't handle backup and restore phase
differently which can be improved.

This patch tells the kernel that the data won't be reused in the copy
function, thus it can discard it from the page cache after it was read.

Next, it will instead tell the kernel that the accused file and the tmp
file both will be needed for reading next. The first call to WILLNEED
should be put up further in the call chain to also support reading the
allocation map but this would invole tracking DONTNEED differently, so
it is not implemented here.

In the end, the page caches will be discarded explicitly by telling the
kernel that the data is no longer needed.

Testing shows that the active page cache during shaking now is released
after each file operation and system responsiveness improves during
shaking.

kakra committed:
Add support for FIEMAP

This patch adds support for FIEMAP and leaves the current FIBMAP
approach in place as a fallback. Instead of trying to guess extents from
FIBMAP block maps, it directly uses extents as fragments.

The logic is similar as it merges adjecent extents as one fragment. It
also detects hole extents and inlined extents and doesn't account those
as fragments similar to the FIBMAP approach.

It optimizes handling filesystems with tail support (e.g., reiserfs) by
not accounting tailed extents as fragments, as the rewrite process
probably would tail the last extent again.

Similarily it tries to optimize for file systems that support shared
extents (xfs, btrfs) by not accounting such extents as fragments. This
helps preventing unsharing accused files and thus increasing space
usage. I pretend that keeping shared extents is more important than
continuous extents, as the operation of deduplicating those extents in
the first place is known to fragment those files.

However, if the heuristics decide to shake the file, extents still
become unshared. I have no idea currently how to optimize for that, thus
it is not part of this patch.

kakra committed:
Cleanup whitespace

kakra committed:
Remove unused variable

kakra committed:
Add FICLONE ioctl from kernel 4.5

kakra committed:
Call fdatasync() for more cooperative performance

Running shake on big files puts a high load on the filesystem. Try to
alleviate this a bit by explicitly calling fdatasync() after truncating
and after writing the destination file. This also ensure the data made
it to disk in case of a crash.

kakra committed:
fcopy() should not ftruncate()

ftruncate() is already explicitly called during rewrite phase, so drop
the redundant call from fcopy() and instead explicitly call it during
the backup phase. This also makes fallocate() more effective, because
the second call to truncate may undo the allocation.

kakra committed:
Only fadvise() if copy was successful

kakra committed:
Add a sanity check

During testing, I found some file sizes to be zero after shaking and I
had to restore from backup. Ensure that this doesn't happen and fails an
assertion early before creating a bigger mess.

kakra committed:
Add reflink support

Instead of doing the backup by copying, just try to reflink the accused
file. This makes a clone of the file data without actually copying the
data on filesystems that support it. If it fails, fall back to the old
behavior and don't try to reflink again.

We are not interested in fragmentation of the clone as it is going to be
deleted anyway, and even with fragmentation it is still good enough as
source for the rewrite phase. This saves 50% of IO to speed up the
process.

kakra committed:
Move the verify assertion into final place

kakra committed:
Merge branch 'master' into improve-performance

kakra commented:
This should be rebased when FIEMAP support and/or fs hinting was merged... Leaving as-is until @unbrice decided how to do the fs hinting.

unbrice commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[executive.c, line 49 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko28IO8wpJ479atsQME:-Ko28IO8wpJ479atsQMF:b-owc6wt) ([raw file](https://github.com/unbrice/shake/blob/aa24a1e5b012be3cc740c144b9337d15d79cbd1b/executive.c#L49)):*
> ```C
>   if (-1 == lseek (in_fd, (off_t) 0, SEEK_SET)
>       || -1 == lseek (out_fd, (off_t) 0, SEEK_SET)
>       || -1 == ftruncate (out_fd, (off_t) 0))
> ```

I suggest you leave this: cost is low but it reduces the preconditions for fcopy() (otherwise callee has to know that out_fd is supposed to be of size 0).

---

*[executive.c, line 232 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko26CuthnOWrVp-YzHy:-Ko26CuthnOWrVp-YzHz:bjaf9y2) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L232)):*
> ```C
> shake_reg_backup_phase (struct accused *a, struct law *l)
> {
>   static int can_reflink = 1;
> ```

I suggest not to cache things: many people have multiple filesytems and won't be able to take advantage of your work. Calling the ioctl to get -1 shouldn't be so expensive.

---

*[executive.c, line 235 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko26rCHgZu0JRPwebkc:-Ko26rCHgZu0JRPwebkd:b-764uxh) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L235)):*
> ```C
>   /* truncate the backup file first, also required for FICLONE to be effective
>    */
>   int res = ftruncate (l->tmpfd, 0);
> ```

Can you add a TODO to use openat(TMP_FILE) instead?

---

*[executive.c, line 242 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko26eta6Ng4_hnJY0TW:-Ko268V6QnIj9qI4ikQD:bhi3pj7) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L242)):*
> ```C
>        * performance
>        */
>       if (can_reflink && (0 == ioctl (l->tmpfd, FICLONE, a->fd)))
> ```

Could you move that FICLONE logic in a separate function in linux.c?

---

*[executive.c, line 255 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko25ZiMcfFW0KmEKsN2:-Ko25ZiMcfFW0KmEKsN3:boiz81o) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L255)):*
> ```C
>           /* give the filesystem a relief
>            */
>           fdatasync (l->tmpfd);
> ```

Are you sure this is useful?

---

*[executive.c, line 268 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko25wrYHjV04PIvFkfj:-Ko25wrYHjV04PIvFkfk:b-yojasn) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L268)):*
> ```C
>               can_reflink = 0;
>               error (0, errno,
>                      "%s: FICLONE failed, falling back to normal copy",
> ```

I don't think we should show anything to the user, as they are not really in a position to act on it anyways.

---

*[executive.c, line 328 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko27bNCD1DacO_N5iMP:-Ko27bNCD1DacO_N5iMQ:b-dli7x7) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L328)):*
> ```C
>     error (1, errno, "%s: restore failed ! file have been saved at %s",
>            a->name, l->tmpname);
>   posix_fadvise (a->fd, (off_t) 0, (off_t) 0, POSIX_FADV_DONTNEED);
> ```

This was merged already, does the push need to be rebased?

---

*[judge.c, line 259 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko27h9EZLHrEtQm9RU1:-Ko27h9EZLHrEtQm9RU2:bhezsw3) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/judge.c#L259)):*
> ```C
>               }
>             else if (z && z->start && labs (z->atime - y->atime) < MAGICTIME)
>               y->ideal = (z->start - y->size - MAGICLEAP);
> ```

Ouch. Thanks.

---

*[linux.c, line 234 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko2A38BO7ZI97hO9wd2:-Ko2A38BO7ZI97hO9wd3:b-377hbe) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/linux.c#L234)):*
> ```C
>   extmap->fm_length = a->size;
>   extmap->fm_flags = FIEMAP_FLAG_SYNC;
>   if (-1 != ioctl (a->fd, FS_IOC_FIEMAP, extmap))
> ```

I suggest making a function like `count_mapped_extents_for_fiemap()` that performs the first call with a first `struct fiemap` on the stack.

---

*[linux.c, line 365 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko28mZonJWNk70Dmmtz:-Ko28mZonJWNk70Dmmu-:b2v3lbz) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/linux.c#L365)):*
> ```C
>       }
>   }
> out:
> ```

I suggest you create two helper functions, one for fiemap,  one for fibmap rather than one huge one with gotos.

---

*[linux.h, line 33 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko2AnNiAVN8NFW4ID8I:-Ko2AnNiAVN8NFW4ID8J:b-i9c4cj) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/linux.h#L33)):*
> ```objc
> 
> # ifndef FICLONE
> #  define FICLONE _IOW(0x94, 9, int) // linux/fs.h since kernel 4.5
> ```

AFAIK, this isn't needed in other modules, wouldn't it be better in the .c file?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/unbrice/shake/14)*
<!-- Sent from Reviewable.io -->

kakra commented:
I'd like to rework and merge this last, after we resolved some of the other open issues. I'll take some of your suggestions to prepare a few separate PRs and rework some of the open PRs.

kakra commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[executive.c, line 255 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko25ZiMcfFW0KmEKsN2:-KuQuNtxfBVyNuVqq9zH:b7itpz9) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L255)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

Are you sure this is useful?
</blockquote></details>

Adding this improves general file system performance while shake is stressing the system. At least on btrfs this feels more responsive. Of course it could be tested tho I'm not sure. Judging file system responsiveness is always subjective. In a non-interactive system it probably reduces speed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/unbrice/shake/14)*
<!-- Sent from Reviewable.io -->

kakra committed:
Remove reflink caching logic

It breaks cross-filesystem portability and the overhead is almost none.

kakra committed:
Add todo to switch to openat()

kakra committed:
Fix one indentation

Well, it's weird to get right...

kakra committed:
Relocate FICLONE logic into separate function

This makes it a little more easy to see what shake_reg_backup_phase is
doing.

kakra committed:
Do not move ftruncate() out of scope

The callee should not know about the preconditions of fcopy() or
fclone(). So move it back into these two functions.

kakra commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[executive.c, line 49 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko28IO8wpJ479atsQME:-KuR5mhrt-1tWTHNmUat:b19ayfo) ([raw file](https://github.com/unbrice/shake/blob/aa24a1e5b012be3cc740c144b9337d15d79cbd1b/executive.c#L49)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

I suggest you leave this: cost is low but it reduces the preconditions for fcopy() (otherwise callee has to know that out_fd is supposed to be of size 0).
</blockquote></details>

I'll refactor it. I had moved it to the callee to have a single point of ftruncate() in preparation of further file copy/clone operations.

---

*[executive.c, line 232 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko26CuthnOWrVp-YzHy:-KuQxS26we0-Q51cWR2f:b-i1h1lr) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L232)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

I suggest not to cache things: many people have multiple filesytems and won't be able to take advantage of your work. Calling the ioctl to get -1 shouldn't be so expensive.
</blockquote></details>

Going to remove it.

---

*[executive.c, line 235 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko26rCHgZu0JRPwebkc:-KuQxlCaLEo7ItqayT4O:bhy9ot6) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L235)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

Can you add a TODO to use openat(TMP_FILE) instead?
</blockquote></details>

Will do.

---

*[executive.c, line 242 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko26eta6Ng4_hnJY0TW:-KuR3ELlccu42ZIVoBRk:b6qd5vi) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L242)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

Could you move that FICLONE logic in a separate function in linux.c?
</blockquote></details>

Will do, it'll then more resemble what fcopy() does which makes sense.

---

*[executive.c, line 268 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko25wrYHjV04PIvFkfj:-KuQxFSLSfslRymTUdLF:b-i1h1lr) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L268)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

I don't think we should show anything to the user, as they are not really in a position to act on it anyways.
</blockquote></details>

Going to remove it.

---

*[executive.c, line 328 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko27bNCD1DacO_N5iMP:-KuR4HsSNV1lmvPeV1DU:b-fxtljn) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/executive.c#L328)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

This was merged already, does the push need to be rebased?
</blockquote></details>

I'll do a rebase after all discussions are resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/unbrice/shake/14)*
<!-- Sent from Reviewable.io -->

kakra committed:
Move FICLONE definition, only needed in executive.c

kakra committed:
Fix a stray LF

kakra commented:
Review status: 3 of 5 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[judge.c, line 259 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko27h9EZLHrEtQm9RU1:-KuR7_xMXFuviyDXfGPU:b-35bv7h) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/judge.c#L259)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

Ouch. Thanks.
</blockquote></details>

I don't see the difference here...

---

*[linux.h, line 33 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko2AnNiAVN8NFW4ID8I:-KuR8d67--CS2vZhFsjF:b-szq6xk) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/linux.h#L33)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

AFAIK, this isn't needed in other modules, wouldn't it be better in the .c file?
</blockquote></details>

I'm moving it over to executive.c, it's the only user.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/unbrice/shake/14)*
<!-- Sent from Reviewable.io -->

kakra commented:
*[linux.c, line 234 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko2A38BO7ZI97hO9wd2:-KuREWHjRTPyvXqfyVUx:bacxrsc) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/linux.c#L234)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

I suggest making a function like `count_mapped_extents_for_fiemap()` that performs the first call with a first `struct fiemap` on the stack.
</blockquote></details>

I seem to not get the purpose of your request: This is how this IOCTL works: You call it twice. First call gets the number of extent maps. If you don't want to know more, you're done. If you need the extent maps (and we do), you realloc, tell it how many extent maps you want and run IOCTL again. The counting is thus the first invocation of the IOCTL already. The rest is just walking the allocation mapping on disk.

Of course I could walk the extents one by one and request only one map per loop but this doesn't gain much except possibly a very high number of IOCTL calls. Also, the extents may change while we scan it that way. I don't think that memory is an issue here, even with thousands of extents, the allocation should be comparably small.

The loop would be much smaller if I didn't try to find adjacent extents but due to how btrfs organizes compression, this is needed. Btrfs breaks compressed files into extents of 128k. Big compressed files usually end up in number of extents in the 100k's although disk allocation is actually continuous. If this were different, we would only do one single IOCTL and have the number of "fragments" which is all we need to know. But well, btrfs compression... As a side effect I skip some types of extents which I don't want to count for fragment calculation, i.e. reflinked extents (we don't want to purposely break reflinks) or inlined/tailed extents (makes no sense to defragment).

Maybe I should instead add some comments to explain the above circumstances?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/unbrice/shake/14)*
<!-- Sent from Reviewable.io -->

kakra commented:
*[linux.c, line 365 at r2](https://reviewable.io:443/reviews/unbrice/shake/14#-Ko28mZonJWNk70Dmmtz:-KuRIF5ASQksvmcA7QTw:b-947ovo) ([raw file](https://github.com/unbrice/shake/blob/df2e8f99f13b4e4287547db107806f8689862688/linux.c#L365)):*
<details><summary><i>Previously, unbrice (Brice Arnould) wrote…</i></summary><blockquote>

I suggest you create two helper functions, one for fiemap,  one for fibmap rather than one huge one with gotos.
</blockquote></details>

That's a good idea but I defer it until later when I have a little more time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/unbrice/shake/14)*
<!-- Sent from Reviewable.io -->

unbrice commented:
Hey :) I'm doing a spring cleanup of old pull requests, please do reopen if you're later interested in resolving the open threads.
